### PR TITLE
drivers: sensor: Add Atmel SAM QDEC (TC) Driver

### PR DESCRIPTION
**Pull request by jpfaff**
Tested on Atmel SMART SAM E70 Xplained board
Origin: Original

### CODE REVIEW

**Comment by mnkp** [comment]
From the point of view of the existing support for Atmel SAM Family in Zephyr the driver code looks good. The only open question is if SENSOR_POSITION parameter should have a unit. In case of QDEC we measure angular rotation.

@galak, @MaureenHelm could you please review this PR.

**Comment by jpfaff** [comment]
@bogdan-davidoaia could you please review this PR?

**Comment by jpfaff** [comment]
I have updated this PR to CMake and resolved merge conflicts.

**Comment by bogdan-davidoaia** [comment]
@jpfaff sorry for the super late answer. Somehow the initial mails got lost in my inbox and only now did I see this PR since it got some updates.

For POSITION I'd change it to ANGULAR_ROTATION or ANGULAR_POSITION for clarity and express it in radians since we already have angular velocity expresses in radians/second for the gyro channels.

**Comment by jpfaff** [comment]
@bogdan-davidoaia no worries.
Different quadrature encoders have different number of steps per revolution, how would you implement that? Kconfig/DTS? 
I always used rotary encorders, but it seems that there exist linear ones as well (some inkjet printers use them for example). We could stick to rotation anyway and with the steps per revolution config the linear movement can be extracted with some calculus, what do you think?

**Comment by bogdan-davidoaia** [comment]
For now, I think you can leave the steps config in the Kconfig.

Not that familiar with these kind of encoders myself, but if they are different enough I'm ok with having two different configs/channels for the rotary and linear ones.

**Comment by jpfaff** [comment]
Sorry, I don't understand the shippable error:
Warning: Could not extract kernel version
Do you have some advice for that?

**Comment by dbkinder** [comment]
The shippable failure "Warning: Could not extract kernel version" is coming from a step in the documentation build process that's trying to extract the kernel version information from the VERSION file (within the doc/conf.py configuration file) and is failing.  @nashif merged a patch (#5329) for this in the master branch, but it has not been applied to the arm branch.

**Comment by galak** [comment]
> The shippable failure "Warning: Could not extract kernel version" is coming from a step in the documentation build process that's trying to extract the kernel version information from the VERSION file (within the doc/conf.py configuration file) and is failing. @nashif merged a patch (#5329) for this in the master branch, but it has not been applied to the arm branch.

I've updated the arm branch, so should be fixed.

**Review by galak — CHANGES_REQUESTED** [review]
1. I don't think drivers/sensor is the correct place for this.  (not sure what is)
2. We need device tree support to get properties like sam_name & steps_per_unit.
3. Can you split out the pinmap update into its own patch from the driver.

**Comment by mnkp** [comment]
> 1. I don't think drivers/sensor is the correct place for this. (not sure what is)

The quadrature decoder driver indeed doesn't fit too well anywhere in the API. There was however a short discussion on the devel list in August regarding the issue and the proposal was to add it to the sensor API (the other possibility would be to extend counter API).
https://lists.zephyrproject.org/pipermail/zephyr-devel/2017-August/007917.html

> 3. Can you split out the pinmap update into its own patch from the driver.

So far we've always kept pinmap in the same commit as the driver. It can be considered part of the SoC support. I just want to mention that having a separate commit for pinmap means change of style. 

On a related note, should we provide pinmap also for sam4s, sam3x series?

File: include/sensor.h [hunk 1]
```diff
@@ -112,13 +112,14 @@ enum sensor_channel {
...
 	/** Altitude, in meters */
 	SENSOR_CHAN_ALTITUDE,
-
 	/** 1.0 micro-meters Particulate Matter, in ug/m^3 */
 	SENSOR_CHAN_PM_1_0,
 	/** 2.5 micro-meters Particulate Matter, in ug/m^3 */
 	SENSOR_CHAN_PM_2_5,
 	/** 10 micro-meters Particulate Matter, in ug/m^3 */
 	SENSOR_CHAN_PM_10,
+	/** Angular position in radians */
```
Comment (mnkp) [inline include/sensor.h]: Is this parameter supposed to change in the range 0 to 2π, -π to π? Is the range unlimited? We probably should clarify this in comment.

**Review by mnkp — COMMENTED** [review]
I have some extra comments (assuming that we want to add this driver to the sensor API and not to the counter API).

**Comment by jpfaff** [comment]
@mnkp I agree with all your comments. 
What can we do to find an agreement on the best place for this kind of drivers?
- In sam e70 qdec is included in the counter module.
- Other socs may have specialized modules for it.
- Until now drivers/sensor is used for (soc)-external components.

**Comment by nashif** [comment]
> Until now drivers/sensor is used for (soc)-external components.

not really, lots o the sensors in drivers/sensors are part of the SoC.

If you are using the sensor.h, then that belongs into sensors.

**Comment by nashif** [comment]
what is the story with this PR? Is this going to be updated and is it still required? If nothing happens within the next week, it will be closed.

**Comment by pdunaj** [comment]
Hi, will anything happen to this PR soon? We need to use other QDEC and I don't want to generate code conflicts or implement things already done?

**Comment by jpfaff** [comment]
Hi,
Not from my side. The project I did this for is dead and I personally have
no time for it.
Take whatever you can use. If needed I can delete the PR, just let me know.

**Comment by pdunaj** [comment]
Thanks, I will try to reuse the API since it was already discussed.

**Comment by pdunaj** [comment]
Hi @jpfaff , you can delete the PR now to avoid any conflict in the future. I will create and issue and PR with our QDEC.

**Pull request closed without merging**